src/main/java/com/limechain/trie/structure/NodeHandle.java

```diff
@@ -0,0 +1,115 @@
+ *     <li>a {@link BranchNodeHandle}, meaning the node is a purely structural branch node with no storage value.</li>
+ * </ul>
+ */
+public abstract sealed class NodeHandle<T> extends Entry<T> permits StorageNodeHandle, BranchNodeHandle {
+    protected final int rawNodeIndex;
+
+    protected NodeHandle(TrieStructure<T> trieStructure, int rawNodeIndex) {
+        super(trieStructure);
+        this.rawNodeIndex = rawNodeIndex;
+    }
+
+    //NOTE:
+    // Later on, we might switch from optionals to nullable return types
+    // Depends on whether Optionals will be more helpful or more annoying
+    /**
+     * Gets the node handle of this node's child at the given index.
+     * @param index The Nibble index (a.k.a. child index) within this node's children.
+     * @return the optional node handle of this child node, empty is no child exists at the given index
+     */
+    public Optional<NodeHandle<T>> getChild(Nibble index) {
+        TrieNode<T> node = this.trieStructure.getNodeAtIndexInner(this.rawNodeIndex);
+        return Optional
+            .ofNullable(node.childrenIndices[index.asInt()])
+            .map(this.trieStructure::nodeHandleAtIndexInner);
+    }
+
+    /**
+     * @return the partial key of the node.
+     */
+    @NotNull
+    public Nibbles getPartialKey() {
+        return this.trieStructure.getNodeAtIndexInner(this.rawNodeIndex).partialKey;
+    }
+
+    /**
+     * @return the full key of the node.
+     */
+    @NotNull
+    public Nibbles getFullKey() {
+        return this.trieStructure.nodeFullKeyAtIndexInner(this.rawNodeIndex);
+    }
+
+    /**
+     * @return the {@link TrieNode#userData} of the underlying node this handle points to.
+     */
+    @Nullable
+    public T getUserData() {
+        return this.trieStructure.getNodeAtIndexInner(this.rawNodeIndex).userData;
+    }
+
+    /**
+     * Sets the {@link TrieNode#userData} of the underlying node, pointed to by this handle.
+     * @param userData the new user data to be set
+     */
+    public void setUserData(@Nullable T userData) {
+        this.trieStructure.getNodeAtIndexInner(this.rawNodeIndex).userData = userData;
+    }
+
+    /**
+     * @return true if this node is the root node of the trie; false otherwise.
+     */
+    public boolean isRootNode() {
+        return Integer.valueOf(this.rawNodeIndex).equals(this.trieStructure.rootIndex);
+    }
+
+    /**
+     * @return the TrieNodeIndex of the node for that handle
+     */
+    public TrieNodeIndex getNodeIndex() {
+        return new TrieNodeIndex(this.rawNodeIndex);
+    }
+
+    /**
+     * @return true if the node has a storage value
+     */
+    public boolean hasStorageValue() {
+        // NOTE: We could do away with a simple `instanceof` check, but this is more explicit
+        return switch (this) {
+            case StorageNodeHandle<T> ignored -> true;
+            case BranchNodeHandle<T> ignored -> false;
+        };
+    }
```

Comment:
`instanceof` generally indicates bad design. Wouldn't it be possible for this function to be made abstract and to have `StorageNodeHandle` and the branch one override it?

Reply:
Yes, we totally could. Let's say that's the Rust influence taking over (plus the design around sealed classes allows for this alternative). Do we want to generally switch to abstract methods? 
- If so, this would apply to a few other places, too. In general, I tend to like this approach more, since abstract methods tend to add indirection (at the very least while inspecting code), and in this particular case: no sensible user would ask the concrete child class `StorageNodeHandle` if it has a storage value - it's kind of obvious, that's its whole point. 😄 So this method only makes sense for the parent class `NodeHandle`.
- But if it's only for this particular occurence (in order to avoid `instanceof` / pattern matching on `this`), I find a good reason for the abstract method.

Reply:
That's generally true, but keep in mind we're going for `sealed class`es in the whole `Entry` hierarchy below, including `NodeHandle`s. So, given a NodeHandle, we can always exhaustively pattern match which actual child class this instance is of (as seen from the method implementation in question). In this case, we lose nothing from the polymorphism, and (what I meant above) I'd even argue we're gaining on minimalism, since we're not introducing any methods into `Storage/BranchNodeHandle` that don't make sense for the respective child class itself (thus adding to their public interfaces what only makes sense for `NodeHandle`'s public interface). 😄
What we lose, well, is this pattern matching which might seem a bit obfuscating (this `ignored` necessity in Java kind of bugs me) depending on perceptions of readability and habits.

Anyway, I just wanted to make sure we're on the same page on that detail. Ultimately, I don't mind either that much (well, maybe I've got a slight bias against abstract methods, especially in such situations where the parent implementor can know it all 😄) given the relatively little relevance of this detail here in particular, but it's a good general discussion.

Just confirm and I'll refactor, unless I've managed to win you over. 😄

P.S. On a separate note, this method (in its current version) should probably be marked `final` as well. 🤔